Follow button

apradillap opened the pull request:
Connects to #842 

### What does this PR do?

- follow a subject:
  - if you click and you are logged in, you start to follow the subject
  - otherwise, the new login modal is opened
- already following a subject
  - if you click, you stop following a subject

Depending on the context, the button should allow the user to follow:

- GobiertoParticipation:
  - a process
  - an issue
  - an event
- GobiertoPeople
  - a person
  - an event


### Pending
- After talking with Alvaro, the notification page has to be redefined because it will grow a lot depending on the people, events, etc. New Issue

![captura de pantalla 2017-10-09 a las 9 47 51](https://user-images.githubusercontent.com/69764/31328764-049c734c-acd7-11e7-8073-2a6d8a8b17eb.png)

apradillap committed:
Add yield follow button to process and issue

apradillap committed:
Add yield follow button to event

apradillap committed:
WIP - Participation alerts

apradillap committed:
Fix trackable process

apradillap committed:
Add tracking into issues

apradillap committed:
Check user_signed_in to show follow button

apradillap committed:
Normalize i18n keys

apradillap committed:
Merge branch 'master' into 842-follow-button

apradillap committed:
Add process and issue in user subscriptions

apradillap committed:
Change subscridable with current_process

apradillap committed:
Fix subscribe and unsuscribe to process and issues

apradillap committed:
Undo modify subscriptions

apradillap committed:
Renamed views and i18n keys

apradillap committed:
Merge branch 'master' into 842-follow-button

apradillap committed:
Complete new follow event and new follow person

apradillap commented:
@Crashillo I have a problem with Follow button, for example in,  http://participacion.gobify.net/participacion/p/pacto-social-mujer
![captura de pantalla 2017-10-09 a las 9 47 51](https://user-images.githubusercontent.com/69764/31345198-455d408a-ad15-11e7-8dfc-f8403148ee04.png)

Can you help me with the magic of CSS? 😄

apradillap commented:
I'll check what happens in staging by following a process, which shows another alert. For example: http://participacion.gobify.net/participacion/p/jjnlvcebnogn-npp0vusbnoirsvon-vlgdlsjbn

furilo committed:
Fix CSS for button

furilo commented:
There you have your CSS. @Crashillo no need to worry.

apradillap committed:
Adapt test to follow button

apradillap commented:
I have tried the following tests:
GobiertoParticipation:
- a process: http://participacion.gobify.net/participacion/p/presupuestos-participativos-alcobendas
- an issue: http://participacion.gobify.net/participacion/temas/cultura
- an event: http://participacion.gobify.net/participacion/p/pacto-social-mujer/agendas/2017-09-05-presentacion-del-pacto-social-por-el-fin-de-la-violencia-contra-la-mujer

GobiertoPeople
- a person: http://participacion.gobify.net/personas/alcalde-de-madrid
- an event: http://participacion.gobify.net/agendas/alcalde-de-madrid/2017-10-26-presentacion-de-presupuestos

@furilo, I have a design detail pending. In "Agendas Module" (Ex. http://participacion.gobify.net/personas/alcalde-de-madrid) the follow button is not prepared for this module with the styles that it has, I pass you a capture. When you follow a person or an event, you don't see the text.

![captura de pantalla 2017-10-10 a las 12 12 38](https://user-images.githubusercontent.com/69764/31381280-55d8bce2-adb4-11e7-85af-9f8faa5c0d12.png)

furilo committed:
Merge branch 'master' into 842-follow-button

furilo committed:
Fix CSS for follow button

furilo commented:
@apradillap the CSS issue is fixed. 

Btw, the idea was that this button was Ajax (in the sandbox you even had the interaction), but now the whole page is reloading. Any special reason?

apradillap commented:
Thanks @furilo for the CSS! I reused another part of the application concerning subscriptions. So I change it to AJAX?

furilo commented:
yes, it its quick

apradillap committed:
Add ajax to subscribable button

ferblape commented:
I think the new styles broke the button About this process:

![screen shot 2017-10-11 at 08 26 37](https://user-images.githubusercontent.com/17616/31425093-077fa204-ae5e-11e7-8680-c620e0a87bd7.png)


See in staging: http://participacion.gobify.net/participacion/p/presupuestos-participativos-alcobendas

ferblape commented on app/controllers/user/subscriptions_controller.rb:
```diff
@@ -38,15 +38,16 @@ def create
     end
 
-    redirect_to request.referrer
+    redirect_to request.referrer if @user_subscription_form.subscribable.class.name == "Site" ||
```
This piece of code is not working well when I follow a Process, because is raising a 500 error:

```
NoMethodError (undefined method `name' for #<GobiertoParticipation::Process:0x007fd7f6ec9ed0>):

app/controllers/user/subscriptions_controller.rb:42:in `create'
```

If you are dealing with HTML and JS requests you should implement a `respond_to`, and probably redirect to the referer in **all** HTML requests, and in the JS requests render a create.js.erb partial. In this partial you should render the button in the new state.

So, when a user follows something the button will say "Proceso seguido!"

ferblape replied:
https://rollbar.com/Populate/gobierto/items/559/

apradillap replied:
This is the mistake I was telling you yesterday that I had pending, I was aware, I correct it and add the test.

ferblape reviewed (changes requested):
I miss tests to  follow a Process. Those tests will have shown you the error 500 I caught in the acceptance.

ferblape commented:
Why I can't see in staging the block Participation in my subscription preferences?

![screen shot 2017-10-11 at 08 44 08](https://user-images.githubusercontent.com/17616/31425655-7b3df4dc-ae60-11e7-832a-239cde5ab4e6.png)

In which branch is it?

apradillap commented:
Alvaro commented that we have to redefine the page of subscriptions, which made no sense because it was going to grow a lot.

apradillap committed:
Add tests and change follow button to respond to js

apradillap committed:
Fix subscription block process event

apradillap commented:
@furilo Except the new styles broke the button About this process that @ferblape comments, I've added tests and I tested:
GobiertoParticipation:
- a process
- an issue

GobiertoPeople
- a person
- an event

I have deleted the follow event button within a process.

amiedes commented on app/forms/gobierto_admin/gobierto_participation/process_form.rb:
```diff
@@ -27,6 +28,14 @@ class ProcessForm
 
       validates :site, :title_translations, :process_type, presence: true
       validates :process_type, inclusion: { in: ::GobiertoParticipation::Process.process_types }
+      validates_presence_of :starts, :ends, if: :process?
```
I wouldn't add these validations over `starts` and `ends`, since they were removed on purpose [here](https://github.com/PopulateTools/gobierto/pull/973) to make the process creation flow less restrictive. Applies to both processes and groups.

Please check with @furilo or @ferblape.

ferblape replied:
Yes, those validations should be removed

ferblape commented on app/views/layouts/application.html.erb:
```diff
@@ -115,11 +115,9 @@
-            <a href="#" class="button outline rounded desktop_only follow_process"><i class="fa fa-rss"></i> <span><%= t('.follow_process') %></span></a>
-          <% end %>
-          <!-- ./ -->
+          <span id="subscribable_button">
```
👍 🆒

apradillap committed:
Remove processes and groups form validations

apradillap committed:
Fix process form test without starts_at and ends_at validations

apradillap commented:
Ready to merge

apradillap committed:
Merge branch 'master' into 842-follow-button
